srml/support/src/storage/hashed/generator.rs

```diff
@@ -203,15 +208,45 @@ pub trait StorageValue<T: codec::Codec> {
 	///
 	/// `T` is required to implement `codec::EncodeAppend`.
 	fn append<S: HashedStorage<Twox128>, I: codec::Encode>(
-		items: &[I], storage: &mut S
-	) -> Result<(), &'static str> where T: codec::EncodeAppend<Item=I> {
+		items: &[I],
+		storage: &mut S,
+	) -> Result<(), &'static str>
+		where T: codec::EncodeAppend<Item=I>
+	{
 		let new_val = <T as codec::EncodeAppend>::append(
 			storage.get_raw(Self::key()).unwrap_or_default(),
 			items,
 		).map_err(|_| "Could not append given item")?;
 		storage.put_raw(Self::key(), &new_val);
 		Ok(())
 	}
+
+	/// Safely append the given items to the value in the storage.
+	///
+	/// `T` is required to implement `codec::EncodeAppend`.
+	fn safe_append<'a, S: HashedStorage<Twox128>, I>(
+		items: &'a[I],
+		storage: &mut S,
+	) where
+		T: codec::EncodeAppend<Item=I> + From<&'a[I]>,
+		I: 'a + codec::Encode + Clone,
+	{
+		Self::append(items, storage).unwrap_or_else(|_| Self::put(&items.clone().into(), storage));
+	}
+
+	/// Read the length of the value in a fast way, without decoding the entire value.
+	///
+	/// `T` is required to implement `Codec::DecodeLength`.
+	fn decode_len<S: HashedStorage<Twox128>>(storage: &mut S) -> Result<usize, &'static str>
+		where T: codec::DecodeLength, T: Len
+	{
+		// attempt to get the length directly.
+		if let Some(k) = storage.get_raw(Self::key()) {
+			<T as codec::DecodeLength>::len(&k).map_err(|e| e.what())
+		} else {
+			Ok(Self::Default::default().len())
```

Comment:
note: this Len trait could have been avoided using: `DecodeLength::len(Self::Default::default().encode())`

but it's probably better this way

Reply:
Not sure which one is more efficient here and makes the most sense. One problem with using an iterator to infer the size that I circumvented but still exists is that Options, including `None`, have size 1 when `into_iter()`-ed which I find weird and can lead to confusion.

https://github.com/paritytech/substrate/blob/346be4baad9286cf2c4f5ae835f6895664383737/srml/support/src/storage/storage_items.rs#L901-L928

Ideally, I would personally do neither and have the codec manually implement an auxiliary `.default_len()` through a trait for all the types that also have `DecodeLength`.

Reply:
I don't understand which iterator you are talking about, I was saying you could encode default value and call DecodeLength on it.

Reply:
hmm also after all I'm not sure I see the issue with using iterator to infer the size: DecodeLength is not implemented on Option so there is no issue there and if it was that would be fine as well to me actually.